apps/prairielearn/src/tests/lti13.test.ts

```diff
@@ -161,155 +318,67 @@ describe('LTI 1.3', () => {
-        type: ['http://purl.imsglobal.org/vocab/lis/v2/course#CourseOffering'],
-        label: 'TEST 101',
-        title: 'Test Course',
-      },
-      name: 'Test User',
-      email: 'test-user@example.com',
-      'https://purl.imsglobal.org/spec/lti/claim/custom': {
+    const executor = await makeLoginExecutor({
+      user: {
+        name: 'Test User',
+        email: 'test-user@example.com',
         uin: '123456789',
+        sub: USER_SUB,
       },
-      'https://purl.imsglobal.org/spec/lti-ags/claim/endpoint': {
-        scope: [
-          'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem',
-          'https://purl.imsglobal.org/spec/lti-ags/scope/score',
-        ],
-        errors: {
-          errors: {},
-        },
-        lineitems: `https://localhost:${oidcProviderPort}/api/lti/courses/1/line_items`,
-        validation_context: null,
-      },
-    })
-      .setProtectedHeader({ alg: 'RS256' })
-      .setIssuer(`http://localhost:${oidcProviderPort}`)
-      .setIssuedAt()
-      .setExpirationTime('1h')
-      .setSubject('a555090c-8355-4b58-b315-247612cc22f0')
-      .setAudience(CLIENT_ID)
-      .sign(joseKey);
-
-    // Run a server to respond to JWKS requests.
-    const app = express();
-    app.get('/jwks', (req, res) => {
-      res.setHeader('Content-Type', 'application/json');
-      // Pass `false` to `toJSON` to only include public keys.
-      res.end(JSON.stringify(keystore.toJSON(false)));
+      fetchWithCookies,
+      oidcProviderPort,
+      keystore,
+      loginUrl: `${siteUrl}/pl/lti13_instance/1/auth/login`,
+      callbackUrl: `${siteUrl}/pl/lti13_instance/1/auth/callback`,
+      targetLinkUri: `${siteUrl}/pl/lti13_instance/1/course_navigation`,
     });
 
-    const finishLoginResponse = await withServer(app, oidcProviderPort, async () => {
-      return await fetchWithCookies(redirectUri, {
-        method: 'POST',
-        body: new URLSearchParams({
-          nonce,
-          state,
-          id_token: fakeIdToken,
-        }),
-      });
-    });
+    const res = await executor.login();
+    assert.equal(res.status, 200);
 
-    assert.equal(finishLoginResponse.status, 200);
     // Inspect more into this response for output that is for Instructors vs for Students
     // Setup link to a course instance (database modification)
     // Confirm the redirect passes through to the course
 
-    const repeatLoginTestNonce = await withServer(app, oidcProviderPort, async () => {
-      return await fetchWithCookies(redirectUri, {
-        method: 'POST',
-        body: new URLSearchParams({
-          nonce,
-          state,
-          id_token: fakeIdToken,
-        }),
-        redirect: 'manual',
-      });
-    });
-
-    // This should fail for nonce reuse
-    assert.equal(repeatLoginTestNonce.status, 500);
+    // Logging in again should fail because of nonce reuse.
+    const repeatLoginRes = await executor.login();
+    assert.equal(repeatLoginRes.status, 500);
```

Comment:
Is this nonce checking what errors with 
`error: Error page {"id":"G38ARCGFGW0H","msg":"did not find expected authorization request details in session, req.session[\"oidc:localhost\"] is undefined"`

Reply:
Yes: https://github.com/PrairieLearn/PrairieLearn/blob/984dc62b2bd67868670cd71f8683af3d6c2c40f9/apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts#L182-L185